### ISSUE TITLE
Add support for playlist cover art

### ISSUE THREAD

**Issue by ForgetfulMagnet** [navidrome/navidrome#406]
Being able to browse a list of playlists is a far better user experience if the playlists have cover art. I believe Airsonic achieves this by using whatever cover.jpg is placed alongside the playlist source, which works fine. Being able to manually choose the image would be even better.

**Comment by deluan** [comment]
I think Airsonic generates the cover based on the album art from the songs that are added to the playlist. I could not find a way to change the playlist artwork manually...

Replicating Airsonic behavior is a good start, I think. I agree that there should be a way to manually select an image from the UI, I may add this in a second moment. I'd love to use [this artwork generator](https://replacecover.com/). If only they had an API....

**Comment by ForgetfulMagnet** [comment]
> I think Airsonic generates the cover based on the album art from the songs that are added to the playlist. I could not find a way to change the playlist artwork manually...
> 
That might be the case if no cover.jpg is available. However I can manually change playlist cover art by changing the jpg to be whatever I like.

**Comment by deluan** [comment]
> That might be the case if no cover.jpg is available. However I can manually change playlist cover art by changing the jpg to be whatever I like.

Where exactly this `cover.jpg` would go? Alongside the song files or in the folder where the `.m3u` file is placed?

**Comment by ForgetfulMagnet** [comment]
Same placement as the tracks.

**Comment by arg274** [comment]
Not sure how that would work if `cover.jpg` is already being used for the album cover/same track is in multiple playlists.

**Comment by TheWrathKing** [comment]
What about a jpg file in the same folder as automatically imported playlists (PlaylistsPath) with the name of the playlist.
For example 'Mariusz Duda - Acoustic Ballads.jpg' for  'Mariusz Duda - Acoustic Ballads.m3u'
A jpg file with this naming convention could also be used in the same folder for playlists created in the web ui.

**Comment by SenorSmartyPants** [comment]
Or the cover art of a random song in the playlist (it could change on each request), if nothing is set? At least for the subsonic API, a screen full of blue vinyl doesn't help picking a playlist.

**Comment by mradermaxlol** [comment]
Could be a good idea to use some hash function output as a filename. Hashing the playlist title + some fixed but unique fields of a particular playlist (e.g. creation date/time or something else) would be enough to have:
1) a fixed name format for a coverart file (a fixed-length sha256 output, for example)
2) no overlapping in filenames of other images

(Now that I think about it, it's probably best to not use the playlist title (as it could be changed by a user) but some technical data instead; something in playlist metadata that doesn't really change after initial creation)

The resulting coverart file could then be placed in some special directory in the music library/Navidrome data dir/etc. Coverart images could then be assigned to the playlists they belong to when the Navidrome server is started (all it takes is calculating hashes of fields used in filename forming and setting some internal struct reflecting the playlist <-> filename binding and serving all that to the client). Actually, the coverart image could even be base64'd and stored in a db file, but I think that might be a bit overkill (and it would also make coverart changing needlessly complicated).

I have no idea how messed up the Subsonic API is in regards to playlist coverarts or how well could this idea fit in with Navidrome's internals, but I hope this helps nevertheless.

**Comment by certuna** [comment]
I don't think it needs to be that complex, the playlist has a UUID in the database as unique identifier. You can see it in the URL: `https://music.mydomain.com/app/#/playlist/63f341f7-33f0-4538-b31f-783c441315ab/show`

**Comment by mradermaxlol** [comment]
Oh, I completely missed the UUID thing. Then, yeah, it would make sense to do it the way you described :) Perhaps with an addition of some coverart storage dir, as cache is unreliable (it can be purged from time to time) and data dir could have little to no available space.

Actually, something like a ./.cover_arts in the medialib directory sounds reasonable.

**Comment by TheWrathKing** [comment]
The UUID thing is indeed unique but if you inadvertently or purposely delete your Navidrome DB and do a full scan again, it is unlikely that auto-imported playlists (and even the ones created in the ui) will be assigned the same UUID again.
The only thing that remains unchanged is the name of the playlist (which is normally also unique), so a jpeg file with the name of the playlist seems to me a better solution.

**Comment by mradermaxlol** [comment]
IMHO using the playlist name is a no-go as there is too big of a name collision probability. And if the database is deleted... Well, it is kind of implied that you lose your settings/data/etc (to some degree at least), which also includes Navidrome-managed coverarts.

Having a human-readable internal coverart name is nice, of course, but do regular users browse the internal Navidrome config/data dir that often? That is kinda the server administrator's job, and that guy is ought to know what he's doing.

Regular users need the ability to change/delete/open/download the coverart, that's about it.

**Comment by deluan** [comment]
> Regular users need the ability to change/delete/open/download the coverart, that's about it.

I agree with that. I think the solution needs to be:
1) If the playlists does not have a cover, generates one based on the cover of the songs (a 4x4 tile is a common approach here)
2) Allow the user to upload/delete the cover from the UI, similar to everything else, including Subsonic
3) Optionally we could allow imported .m3u's to also import a cover if it is present alongside the m3u itself (same name, extension .png or .jpg), or using an [extended tag](https://en.wikipedia.org/wiki/M3U#Extended_M3U): `#EXTINF:123 logo="cover.jpg",Track Title`. Not sure about this, though, as it is not standard

**Comment by TheWrathKing** [comment]
I totally agree with the possibility to provide a cover along with an imported playlist (same name with jpg or png extension).
The possibility to upload a cover in the ui would also be a great addition.
And by the way, thank you very much for this wonderful piece of software.

**Comment by metalheim** [comment]
@deluan 1. (generating a 2x2 colalge from 4 of the playlist's songs covers) is rather straightforwad to do.
It can be done using ffmpeg without the need for any new dependency and neatly ffmpeg can handle virtually any image format that one might use for cover images.

**This command takes 4 images and outputs a 2x2 collage from the inputs.** 

`ffmpeg \
-i cover1.jpg -i cover2.bmp -i cover3.webp -i cover4.png \
-filter_complex \
"[0][1]scale2ref[scaled0][scaled1]; \
[scaled1][scaled0]hstack=inputs=2[top]; \
[2][3]scale2ref[scaled2][scaled3]; \
[scaled2][scaled3]hstack=inputs=2[bottom]; \
[top][bottom]scale2ref[topscaled][bottomscaled]; \
[topscaled][bottomscaled]vstack=inputs=2[collage]; \
[collage]scale=300:300[collage-resized]" \
-map "[collage-resized]" \
collage_resized.jpg`

It scales all images to match different input sizes and aspect ratios and outputs a compressed 300x300 pixel collage. 
This basically "just" needs a wrapper function that generates it on request and hooks into the caching logic, so it doesnt need to be generated everytime.

**Comment by copart** [comment]
Would be great if nsp's had a header that could include the cover url.  However, my vote would be whatever makes it easiest to implement. :)

**Comment by TheWrathKing** [comment]
Any news about this possibility ?
"Optionally we could allow imported .m3u's to also import a cover if it is present alongside the m3u itself (same name, extension .png or .jpg"

**Comment by HipyCas** [comment]
Any updates on this? I am using an app in my android phone with my Navidrome instance and would really like to set some covers for my playlists

**Comment by pto199** [comment]
I find it strange that this was never implemented.  Do people not listen to playlists?  I have like 80 playlists and it sucks that they all show no cover, when I have covers even for all of them.

**Comment by JaxzKyd** [comment]
> I find it strange that this was never implemented. Do people not listen to playlists? I have like 80 playlists and it sucks that they all show no cover, when I have covers even for all of them.

Me as well! Hard to believe this has been sitting here for 2 years

**Comment by deluan** [comment]
> I find it strange that this was never implemented. Do people not listen to playlists? I have like 80 playlists and it sucks that they all show no cover, when I have covers even for all of them.

I do listen to playlists, and I agree that it sucks to not have covers on them. But it also sucks that the search is slow for my big library, it also sucks the transcoding options are not great, it also sucks there's no support for multiple artists, and multiple folders, and gapless playback, etc, etc... And on top of it there's my day job (you know, to pay the bills and stuff). 

I try my best to dedicate a good chunk of my free time to work on side projects (including the main one, Navidrome), but there's not enough hours in a day to do all i want.

>  any plans on implementing this anymore, or is it unplanned?

If it is an open issue, means that I'll get to it eventually. **As always, PRs are mostly welcome.** ™️

**Comment by Szeraax** [comment]
@deluan Do you want to add this as planned work for the react v5 rewrite? Or sometime after?

**Comment by ShadowsDieThrice** [comment]
I have a working setup with custom playlist art, by making some relatively simple modifications to the /core/artwork/reader_playlist.go file. The modifications add a couple of functions to check a sub-directory named "zPlaylists" in my music folder, and matches the playlist name to the image file name within said folder, before falling back to grid generation/placeholder image if a corresponding image file is not found. But I can quite easily tweak this to accommodate just using the root of the music folder or other setups.

Am happy to open a PR if there's still interest - but I'm a little unfamiliar with this sort of etiquette, sorry!

Changes are in a52d7a3c6ddfbff2573e92b853ac439de966525d

**Comment by TheWrathKing** [comment]
Hello @deluan , now that playlists artworks are visible in the webUI, it would be great to have the possibility to choose from custom coverarts (image with the same name as auto-imported m3u or nsp files).

Maybe a new configuration option, for example PlaylistArtPriority, with a default value of "4x4, filename"

4x4 collages are fine in the webUI but on mobile phones it's quite difficult to differentiate playlists because of the small screen. A custom playlist artwork would make things easier.

Thanks a lot for the great work and the constant improvement to this music server.

**Comment by deluan** [comment]
There is some work done in #3727, I'll try to finish it up for next release

**Comment by motoyugota** [comment]
Any update on this?

**Comment by adrbn** [comment]
I would be curious to know too whether this option would be feasible in the near future? Thank you again for your time and work, Navidrome is absolutely amazing!

**Comment by mvastum** [comment]
Just keeping this alive, I'd also be interested in this feature. The auto-created covers look all the same to me. The iOS App Narjo has this feature implemented already, but it's not a server-side solution.
